python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+ *    flow step.
+ *
+ * 1. [Read] Content is read from `TIterableSequence(sequence)` into  `TIterableElement(sequence)`.
+ *    If `sequence` is of type tuple, we will not read tuple content as that would allow
+ *    cross talk.
+ *
+ * 1. [Store] Content is stored from `TIterableElement(sequence)` to `sequence`.
+ *    Here the content type is chosen according to the type of sequence.
+ *
+ * 1. [Read] Content is read from `sequence` to its elements according to the type of `sequence`.
+ *    If the element is a plain variable, the target is the corresponding essa node.
+ *    If the element is itelf a sequence, with control-flow node `seq`, the target is `TIterableSequence(seq)`.
+ *    If the element is a starred variable, with control-flow node `v`, the target is `TIterableElement(v)`.
+ *
+ * 1. [Store] Content is stored from `TIterableElement(v)` to the essa variable for `v`, with
+ *    content type `ListElement`.
+ *
+ * 1. [Flow, Read, Store] The last 5 steps are repeated for all recursive elements which are sequences.
+ */
+module UnpackingAssignment {
+  /** A direct (or top-level) target of an unpacking assignment */
+  class UnpackingAssignmentDirectTarget extends ControlFlowNode {
+    Expr value;
+
+    UnpackingAssignmentDirectTarget() {
+      this instanceof SequenceNode and
+      exists(Assign assign | this.getNode() = assign.getATarget() | value = assign.getValue())
+    }
+
+    Expr getValue() { result = value }
+  }
+
+  /** A (possibly recursive) target of an unpacking assignment */
+  class UnpackingAssignmentTarget extends ControlFlowNode {
+    UnpackingAssignmentTarget() {
+      this instanceof UnpackingAssignmentDirectTarget
+      or
+      this = any(UnpackingAssignmentTarget parent).getAnElement()
+    }
+
+    ControlFlowNode getElement(int i) { result = this.(SequenceNode).getElement(i) }
+
+    ControlFlowNode getAnElement() { result = this.getElement(_) }
+  }
+
+  predicate unpackingAssignmentFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(UnpackingAssignmentTarget target | target instanceof SequenceNode |
+      nodeFrom = TIterableSequence(target) and
+      nodeTo.asCfgNode() = target
+    )
+  }
+
+  predicate unpackingAssignmentConvertingReadStep(Node nodeFrom, Content c, Node nodeTo) {
+    exists(UnpackingAssignmentTarget target | target instanceof SequenceNode |
+      nodeFrom = TIterableSequence(target) and
+      nodeTo = TIterableElement(target) and
+      (
+        c instanceof ListElementContent
+        or
+        c instanceof SetElementContent
+        or
+        // do not lose precision by routing tuple content through the `IterableElement`
+        not target instanceof TupleNode and
+        // `index` refers to `nodeFrom`, but only the ones in `target` are relevant.
+        exists(int index | exists(target.getElement(index)) |
+          c.(TupleElementContent).getIndex() = index
+        )
+        // leaving out dict content for now
```

Comment:
The thing is, it would unpack the keys, so I am not sure how to handle that. But I agree that we should either make it a TODO or write in a more prominent place that we are not handling it.